src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -0,0 +1,143 @@
+# -*- coding: utf-8 -*-
+#
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import warnings
+from typing import Optional, TypeGuard
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator, SymbolTableTrait
+from functional.common import GTSyntaxError
+
+
+def is_complete_type(fo_type: foast.SymbolType) -> TypeGuard[foast.SymbolType]:
```

Comment:
Maybe `is_complete_symbol_type()` would be a better name here since the function is now a `SymbolType` typeguard